metagraph/core/planning.py

```diff
@@ -13,24 +13,33 @@ def __init__(self, src_type):
         self.src_type = src_type
         self.translators = []
         self.dst_types = []
+        self.unsatisfiable_dst_type = None
 
     def __len__(self):
+        if self.unsatisfiable_dst_type is not None:
+            raise ValueError(
+                "No translation path found for {src_type.__name__} -> {self.unsatisfiable_dst_type.__name__}"
+            )
         return len(self.translators)
 
     def __iter__(self):
+        if self.unsatisfiable_dst_type is not None:
+            raise ValueError(
+                "No translation path found for {src_type.__name__} -> {self.unsatisfiable_dst_type.__name__}"
+            )
         return iter(self.translators)
 
     def __repr__(self):
-        if self.src_type is None or not self.dst_types:
-            return f"{self.__class__.__name__}"
-        return f"{self.__class__.__name__}({self.src_type.__name__} -> {self.dst_types[-1].__name__})"
-
-    def __str__(self):
-        if len(self) == 0:
-            return "No translation required"
-
         s = []
-        if len(self) > 1:
+        if self.unsatisfiable_dst_type is not None:
+            s.append("[Unsatisfiable Translation]")
+            s.append(
+                "Translation to {self.unsatisfiable_dst_type.__name__} unsatisfiable"
+            )
+        elif len(self) == 0:
+            s.append("[Null Translation]")
+            s.append("No translation required")
```

Comment:
For the unsatisfiable case, can you include the src_type for completeness? Something like "Translation from {self.src_type.__name__} to {self.unsatisfiable_dst_type.__name__} is unsatisfiable"

Same goes for the Null translation case. Mentioning the src and dst types will aid in debugging, especially given this is a case that should not normally happen.

Reply:
That sounds like a good idea. 

https://github.com/metagraph-dev/metagraph/pull/110/commits/50264b6875625c71080b4fdda2f1bd7211dcb262 changes the printing for `[Null Translation]` and `[Unsatisfiable Translation]`.

We now have the following possible results:
```
[Unsatisfiable Translation]
Translation SRC_TYPE -> DST_TYPE unsatisfiable
```
```
[Null Translation]
No translation required from SRC_TYPE to itself
```
```
[Multi-step Translation]
(start)  SRC_TYPE
           -> INTERMEDIATE_TYPE
(end)        -> DST_TYPE
```
```
[Direct Translation]
SRC_TYPE -> DST_TYPE
```